File: Example/PrebidDemo/PrebidDemoObjectiveC/Examples/MAX/MAXNativeViewController.m
/*   Copyright 2019-2022 Prebid.org, Inc.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "MAXNativeViewController.h"
#import "PrebidDemoMacros.h"

NSString * const nativeStoredImpressionMAX = @"prebid-demo-banner-native-styles";
NSString * const maxRenderingNativeAdUnitId = @"240da3ba91611d72";

@interface MAXNativeViewController ()

// Prebid
@property (nonatomic) MediationNativeAdUnit * maxMediationNativeAdUnit;
@property (nonatomic) MAXMediationNativeUtils * mediationDelegate;

// MAX
@property (nonatomic) MANativeAdLoader * maxNativeAdLoader;
@property (nonatomic, weak) MAAd * maxLoadedNativeAd;

@end

@implementation MAXNativeViewController

- (void)loadView {
    [super loadView];
    
    [self createAd];
}

- (void)dealloc {
    if (self.maxLoadedNativeAd != nil) {
        [self.maxNativeAdLoader destroyAd:self.maxLoadedNativeAd];
    }
}

- (void)createAd {
    // 1. Create a MANativeAdLoader
    self.maxNativeAdLoader = [[MANativeAdLoader alloc] initWithAdUnitIdentifier:maxRenderingNativeAdUnitId];
    self.maxNativeAdLoader.nativeAdDelegate = self;
    
    // 2. Create the MAXMediationNativeUtils
    self.mediationDelegate = [[MAXMediationNativeUtils alloc] initWithNativeAdLoader:self.maxNativeAdLoader];
    
    // 3. Create the MediationNativeAdUnit
    self.maxMediationNativeAdUnit = [[MediationNativeAdUnit alloc] initWithConfigId:nativeStoredImpressionMAX mediationDelegate:self.mediationDelegate];
    
    // 4. Configure the MediationNativeAdUnit
    NativeAssetImage *image = [[NativeAssetImage alloc] initWithMinimumWidth:200 minimumHeight:200 required:true];
    image.type = ImageAsset.Main;
    
    NativeAssetImage *icon = [[NativeAssetImage alloc] initWithMinimumWidth:20 minimumHeight:20 required:true];
    icon.type = ImageAsset.Icon;
    
    NativeAssetTitle *title = [[NativeAssetTitle alloc] initWithLength:90 required:true];
    NativeAssetData *body = [[NativeAssetData alloc] initWithType:DataAssetDescription required:true];
    NativeAssetData *cta = [[NativeAssetData alloc] initWithType:DataAssetCtatext required:true];
    NativeAssetData *sponsored = [[NativeAssetData alloc] initWithType:DataAssetSponsored required:true];
    
    NativeEventTracker * eventTracker = [[NativeEventTracker alloc] initWithEvent:EventType.Impression
                                                                          methods:@[EventTracking.Image, EventTracking.js]];
    [self.maxMediationNativeAdUnit addNativeAssets:@[title, icon, image, sponsored, body, cta]];
    [self.maxMediationNativeAdUnit addEventTracker:@[eventTracker]];
    
    [self.maxMediationNativeAdUnit setContextType:ContextType.Social];
    [self.maxMediationNativeAdUnit setPlacementType:PlacementType.FeedContent];
    [self.maxMediationNativeAdUnit setContextSubType:ContextSubType.Social];
    
    // 5. Create a MAXNativeAdView
    UINib * nativeAdViewNib = [UINib nibWithNibName:@"MAXNativeAdView" bundle:NSBundle.mainBundle];
    MANativeAdView * maNativeAdView = [nativeAdViewNib instantiateWithOwner:nil options:nil].firstObject;
    
    // 6. Create a MANativeAdViewBinder
    MANativeAdViewBinder * adViewBinder = [[MANativeAdViewBinder alloc] initWithBuilderBlock:^(MANativeAdViewBinderBuilder * _Nonnull builder) {
        builder.iconImageViewTag = 1;
        builder.titleLabelTag = 2;
        builder.bodyLabelTag = 3;
        builder.advertiserLabelTag = 4;
        builder.callToActionButtonTag = 5;
        builder.mediaContentViewTag = 123;
    }];
    
    // 7. Bind views
    [maNativeAdView bindViewsWithAdViewBinder:adViewBinder];
    
    // 7. Make a bid request to Prebid Server
    @weakify(self);
    [self.maxMediationNativeAdUnit fetchDemandWithCompletion:^(enum ResultCode resultCode) {
        @strongify(self);
        if (!self) { return; }
        
        // 8. Load the native ad
        [self.maxNativeAdLoader loadAdIntoAdView:maNativeAdView];
    }];
}

// MARK: - MANativeAdDelegate

- (void)didLoadNativeAd:(MANativeAdView *)nativeAdView forAd:(MAAd *)ad {
    if (self.maxLoadedNativeAd != nil) {
        [self.maxNativeAdLoader destroyAd:self.maxLoadedNativeAd];
    }
    
    self.maxLoadedNativeAd = ad;
    
    self.bannerView.backgroundColor = [UIColor clearColor];
    
    nativeAdView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.bannerView addSubview:nativeAdView];
    
    [[self.bannerView.heightAnchor constraintEqualToAnchor:nativeAdView.heightAnchor] setActive:YES];
    [[self.bannerView.topAnchor constraintEqualToAnchor:nativeAdView.topAnchor] setActive:YES];
    [[self.bannerView.leftAnchor constraintEqualToAnchor:nativeAdView.leftAnchor] setActive:YES];
    [[self.bannerView.rightAnchor constraintEqualToAnchor:nativeAdView.rightAnchor] setActive:YES];
}

- (void)didFailToLoadNativeAdForAdUnitIdentifier:(NSString *)adUnitIdentifier withError:(MAError *)error {
    PBMLogError(@"%@", error.message);
}

- (void)didClickNativeAd:(MAAd *)ad {
}

@end

File: Example/PrebidDemo/PrebidDemoSwift/Examples/MAX/MAXNativeViewController.swift
/*   Copyright 2019-2022 Prebid.org, Inc.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import UIKit
import PrebidMobile
import PrebidMobileMAXAdapters
import AppLovinSDK

fileprivate let nativeStoredImpression = "prebid-demo-banner-native-styles"
fileprivate let maxRenderingNativeAdUnitId = "e4375fdcc7c5e56c"

class MAXNativeViewController: BannerBaseViewController, MANativeAdDelegate {
    
    // Prebid
    private var maxMediationNativeAdUnit: MediationNativeAdUnit!
    private var maxMediationDelegate: MAXMediationNativeUtils!
    
    private var nativeRequestAssets: [NativeAsset] {
        let image = NativeAssetImage(minimumWidth: 200, minimumHeight: 50, required: true)
        image.type = ImageAsset.Main
        
        let icon = NativeAssetImage(minimumWidth: 20, minimumHeight: 20, required: true)
        icon.type = ImageAsset.Icon
        
        let title = NativeAssetTitle(length: 90, required: true)
        let body = NativeAssetData(type: DataAsset.description, required: true)
        let cta = NativeAssetData(type: DataAsset.ctatext, required: true)
        let sponsored = NativeAssetData(type: DataAsset.sponsored, required: true)
        
        return [title, icon, image, sponsored, body, cta]
    }
    
    private var eventTrackers: [NativeEventTracker] {
        [NativeEventTracker(event: EventType.Impression, methods: [EventTracking.Image,EventTracking.js])]
    }
    
    // MAX
    private var maxNativeAdLoader: MANativeAdLoader!
    private weak var maxLoadedNativeAd: MAAd!
    
    override func loadView() {
        super.loadView()
        
        createAd()
    }
    
    deinit {
        if let maxLoadedNativeAd = maxLoadedNativeAd {
            maxNativeAdLoader?.destroy(maxLoadedNativeAd)
        }
    }
    
    func createAd() {
        // 1. Create a MANativeAdLoader
        maxNativeAdLoader = MANativeAdLoader(adUnitIdentifier: maxRenderingNativeAdUnitId)
        maxNativeAdLoader.nativeAdDelegate = self
        
        // 2. Create the MAXMediationNativeUtils
        maxMediationDelegate = MAXMediationNativeUtils(nativeAdLoader: maxNativeAdLoader)
        
        // 3. Create the MediationNativeAdUnit
        maxMediationNativeAdUnit = MediationNativeAdUnit(configId: nativeStoredImpression, mediationDelegate: maxMediationDelegate)
        
        // 4. Configure the MediationNativeAdUnit
        maxMediationNativeAdUnit.addNativeAssets(nativeRequestAssets)
        maxMediationNativeAdUnit.setContextType(.Social)
        maxMediationNativeAdUnit.setPlacementType(.FeedContent)
        maxMediationNativeAdUnit.setContextSubType(.Social)
        maxMediationNativeAdUnit.addEventTracker(eventTrackers)
        
        // 5. Create a MAXNativeAdView
        let nativeAdViewNib = UINib(nibName: "MAXNativeAdView", bundle: Bundle.main)
        let maNativeAdView = nativeAdViewNib.instantiate(withOwner: nil, options: nil).first! as! MANativeAdView?
        
        // 6. Create a MANativeAdViewBinder
        let adViewBinder = MANativeAdViewBinder.init(builderBlock: { (builder) in
            builder.iconImageViewTag = 1
            builder.titleLabelTag = 2
            builder.bodyLabelTag = 3
            builder.advertiserLabelTag = 4
            builder.callToActionButtonTag = 5
            builder.mediaContentViewTag = 123
        })
        
        // 7. Bind views
        maNativeAdView!.bindViews(with: adViewBinder)
        
        // 7. Make a bid request to Prebid Server
        maxMediationNativeAdUnit.fetchDemand { [weak self] result in
            // 8. Load the native ad
            self?.maxNativeAdLoader.loadAd(into: maNativeAdView!)
        }
    }
    
    // MARK: - MANativeAdDelegate
    
    func didLoadNativeAd(_ nativeAdView: MANativeAdView?, for ad: MAAd) {
        if let nativeAd = maxLoadedNativeAd {
            maxNativeAdLoader?.destroy(nativeAd)
        }
        
        maxLoadedNativeAd = ad
        
        bannerView.backgroundColor = .clear
        nativeAdView?.translatesAutoresizingMaskIntoConstraints = false
        bannerView.addSubview(nativeAdView!)
        
        bannerView.heightAnchor.constraint(equalTo: nativeAdView!.heightAnchor).isActive = true
        bannerView.topAnchor.constraint(equalTo: nativeAdView!.topAnchor).isActive = true
        bannerView.leftAnchor.constraint(equalTo: nativeAdView!.leftAnchor).isActive = true
        bannerView.rightAnchor.constraint(equalTo: nativeAdView!.rightAnchor).isActive = true
    }
    
    func didFailToLoadNativeAd(forAdUnitIdentifier adUnitIdentifier: String, withError error: MAError) {
        PrebidDemoLogger.shared.error("\(error.message)")
    }
    
    func didClickNativeAd(_ ad: MAAd) {}
}
